### DISCUSSION TITLE
How to navigate in DocumentView when developing extensions for VS 2022?

### DISCUSSION THREAD

**Discussion by StanislavKoncebovski** [VsixCommunity/Community.VisualStudio.Toolkit#355] [General]
Mads Kristensen gave an enlightening explanation how to insert a GUID at a current caret position. But literally nowhere was I able to find a hint to how to insert a text at a different defined position in the document, e.g. at the very beginning. I failed to find any comprehensible documentation or tutorial to this topic. I guess, a possibility to navigate must be present, but very well hidden from a layman like myself. I spent a couple of hours trying to understand the classes, interfaces, etc. responsible for this episode by brute debugging, but, so far, failed, and I hope that someone in this universe could provide some explanation. TIA!

**Comment by reduckted** [comment]
The `ITextBuffer` allows you to insert text at arbitrary points. You can also use its `CurrentSnapshot` property to get the length of the document and the lines in the document.

Here are some examples.

```cs
DocumentView document = await VS.Documents.GetActiveDocumentViewAsync();

// Insert at the start of the document.
document.TextBuffer.Insert(0, "I'm at the start.");

// Insert at the end of the document.
int end = document.TextBuffer.CurrentSnapshot.Length;
document.TextBuffer.Insert(end, "I'm at the end.");

// Inserting at the start of the second line.
ITextSnapshotLine secondLine = document.TextBuffer.CurrentSnapshot.GetLineFromLineNumber(1);
document.TextBuffer.Insert(secondLine.Start, "Now I am the second line\n.");
```

**Reply by StanislavKoncebovski** [reply]
Thanks a lot! I tested it, and it works!
Stan

**Reply by StanislavKoncebovski** [reply]
BTW, is there an API documentation? I would be thankful for a hint!

**Reply by reduckted** [reply]
You can find the API here: https://docs.microsoft.com/dotnet/api/microsoft.visualstudio.text.itextbuffer

There's also a good overview of the whole editor here: https://docs.microsoft.com/visualstudio/extensibility/inside-the-editor

**Reply by StanislavKoncebovski** [reply]
Thanks a lot Dave!
